### ISSUE TITLE
std::out_of_range from deque::_M_range_check

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#2143]
We have seen the following streaming failure when trying to decommission a node and having it stream its data to other nodes:

```
stream_transfer_task: Fail to send to ....: std::out_of_range (deque::_M_range_check: __n (which is 64)>= this->size() (which is 64))
```

The deque::_M_range_check exception comes, obviously, from std::deque. I inspected the std::deque header files, and it appears this exception can only come from using the at() function. Inspecting our code (both Seastar and Scylla), the only use of deque::at() is in compress.cc's locate() function.

This locate() must be called with a valid position inside the uncompressed file - the file's total uncompressed size is not a valid position, and trying to pass it to locate() can lead us to looking at one past the last item in the deque if we're unlucky (if the file size is a multiple of the chunk size) and exactly the kind of exception we saw above.

Inspecting the users of locate(), it seems only the skip() function has this bug where it might pass the end of file position to locate().

I still need to prove this skip() bug explains the message we saw. It would require the streaming code to actually use skip(), and also to try skipping to the end of the file at some point, and I still didn't verify that it does. I'm also surprised we got this error message with low numbers like "64" - does this mean we have some very small sstables in that case (only 64 compression chunks)?

**Comment by nyh** [comment]
Note that this issue is similar to #1039 (which involved a different bug in the compress.cc code, resulting in a call to locate(-1)).

**Comment by tgrabiec** [comment]
@nyh streaming uses skip() since 293876c72fd43c88c03f7a922859aa94565944a5. `multi_range_mutation_reader` is using `fast_forward_to()` for each sub-range it is asked to stream. It can happen that for some sstables current sub-range will be after all partitions, in which case we will skip to the end.

I think the past-the-end-of-file position is a valid position, and thus we should allow skipping to it. A pair of positions is like a pair of iterators, and past-the-end is like container's end(). From that perspective, I think locate() should handle this.

**Comment by nyh** [comment]
Thanks Tomek for this explanation!

I agree that past-the-end-of-file is indeed a valid position - the normal reading already supports it (fixed in commit 02ba8ffbe842b369cf4bbdddca18fae0eac36d29) but the relatively-recently-added skip()  supported it incorrectly, and I'll send a patch to fix it.

I don't think the locate() function itself should handle the past-the-end position because it doesn't return an iterator, rather an actual compressed chunk to be read - and in this case there is no compressed chunked to be read, and the caller should return nothing. I guess I could change locate() to return an optional<chunk_and_offset>, but this won't simplify the callers much - as they would still need to check if the returned optional is engaged. Given that locate() is just an internal function used in  3 places in compress.cc, I prefer to fix the callers (and two of them were already correct).

**Comment by asias** [comment]
@nyh is the patch you will send supposed to fix the deque problem?

**Comment by nyh** [comment]
Yes, that's my intention. But I don't yet have a reproducer to test this.

**Comment by tgrabiec** [comment]
2017-03-08 10:11 GMT+01:00 nyh <notifications@github.com>:

> Thanks Tomek for this explanation!
>
> I agree that past-the-end-of-file is indeed a valid position - the normal
> reading already supports it (fixed in commit 02ba8ff
> <https://github.com/scylladb/scylla/commit/02ba8ffbe842b369cf4bbdddca18fae0eac36d29>)
> but the relatively-recently-added skip() supported it incorrectly, and I'll
> send a patch to fix it.
>
> I don't think the locate() function itself should handle the past-the-end
> position because it doesn't return an iterator, rather an actual compressed
> chunk to be read - and in this case there is no compressed chunked to be
> read, and the caller should return nothing. I guess I could change locate()
> to return an optional<chunk_and_offset>, but this won't simplify the
> callers much - as they would still need to check if the returned optional
> is engaged. Given that locate() is just an internal function used in 3
> places in compress.cc, I prefer to fix the callers (and two of them were
> already correct).
>
That should be ok too. Though note that even though visually callers would
not appear simpler, using optional<> could be safer because having to deal
with optional<> forces the caller to deal with this case, while in the
other solution it is easy to forget that past-the-end needs to be handled
(as evident here).

**Comment by tgrabiec** [comment]
btw, I think on master this won't reproduce in the same circumstances because of bcc826cc344215243369f7cadc272afaf7c6024b, which avoids the skip for empty ranges.
